great_ape_safe/ape_api/ebtc.py

```diff
@@ -866,3 +876,277 @@ def authority_set_authority(self, new_authority):
+            self.collateral.balanceOf(self.safe.address)
+        ) == (
+            cdp_id_coll
+            + cdp_id_liquidator_reward_shares
+            + self.collateral.getSharesByPooledEth(collateral_balance_before)
+        )
+
+        # 2.3. verify expected values are 0 at readings from the cdp manager
+        (
+            cdp_id_debt,
+            cdp_id_coll,
+            cdp_id_stake,
+            cdp_id_liq_reward_shares,
+            _,
+            _,
+        ) = self.cdp_manager.Cdps(cdp_id)
+        assert self.cdp_manager.cdpStEthFeePerUnitIndex(cdp_id) == 0
+        assert cdp_id_debt == 0
+        assert cdp_id_coll == 0
+        assert cdp_id_stake == 0
+        assert cdp_id_liq_reward_shares == 0
+
+    def cdp_add_collateral(self, cdp_id, coll_amount):
+        """
+        @dev Adds the received stETH to the specified Cdp.
+        @param cdp_id The CdpId on which this operation is operated.
+        @param coll_amount The total stETH collateral amount deposited (added) for the specified Cdp.
+        """
+        # verify: is it coll balance available
+        self._assert_collateral_balance(coll_amount)
+
+        # verify: cdp id ownership from caller
+        self._assert_cdp_id_ownership(cdp_id)
+
+        feed_price = self.price_feed.fetchPrice.call()
+        prev_icr = self.cdp_manager.getSyncedICR(cdp_id, feed_price)
+        prev_tcr = self.cdp_manager.getSyncedTCR(feed_price)
+        prev_coll_balance = self.cdp_manager.getCdpCollShares(cdp_id)
+
+        # 1. collateral approval for BO
+        self.collateral.approve(self.borrower_operations.address, coll_amount)
+
+        # 2. increase collateral in target cdp id
+        self.borrower_operations.addColl(cdp_id, cdp_id, cdp_id, coll_amount)
+
+        # 3. assertions:
+
+        # 3.1. icr and tcr had increased
+        assert self.cdp_manager.getSyncedICR(cdp_id, feed_price) > prev_icr
+        assert self.cdp_manager.getSyncedTCR(feed_price) > prev_tcr
+
+        # 3.2 collateral in cdp at storage is exact on "shares" terms, following internal cdp accounting
+        # after increase should be equal to final = initial + top-up collateral
+        assert self.cdp_manager.getCdpCollShares(
+            cdp_id
+        ) == prev_coll_balance + self.collateral.getSharesByPooledEth(coll_amount)
+
+    def cdp_withdraw_collateral(self, cdp_id, coll_amount):
+        """
+        @dev Withdraws the amount of collateral from the specified Cdp.
+        @param cdp_id The CdpId on which this operation is operated.
+        @param coll_amount The total stETH collateral amount withdrawn (reduced) for the specified Cdp.
+        """
+        # verify: cdp id ownership from caller
+        self._assert_cdp_id_ownership(cdp_id)
+
+        # verify: cdp holds enough collateral to be withdrawn
+        (
+            _,
+            cdp_id_coll,
+            _,
+            _,
+            _,
+            _,
+        ) = self.cdp_manager.Cdps(cdp_id)
```

Comment:
similarly, please consider to user `cdp_manager.getSyncedDebtAndCollShares(cdp_id)`

Reply:
suggestion added in 4fcce46cf1cb8a85bbef7007a50b6af49de60e10